python/ray/data/dataset.py

```diff
@@ -845,14 +845,128 @@ def groupby(self, key: "GroupKeyT") -> "GroupedDataset[T]":
         from ray.data.grouped_dataset import GroupedDataset
         return GroupedDataset(self, key)
 
-    def aggregate(self, *aggs: AggregateFn) -> U:
+    def agg(self, aggs: Union[str, AggregateFn, List[Union[str, AggregateFn]],
```

Comment:
I think we need to accept `Aggregation` and `List[Aggregation]` as well since it's most expressive: users can provide callable `on` etc.

Reply:
My intention here with accepting plain `AggregateFn` instead of `Aggregation` was [following the Pandas `.agg()` semantics](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.agg.html), where `Callable` or `List[Callable]` is applied to all columns. I.e., for the Pandas `.agg()` API, the structure of the arguments dictate the column(s) that the given aggregations will be applied to:
- `Union[str, AggregateFn, List[Union[str, AggregateFn]]]`: Aggregation(s) are applied to each column.
- `Dict[str, List[Union[str, AggregateFn]]]`: Aggregations are applied to the corresponding key column.

Since it's the structure of the arguments and not the arguments themselves that dictate the columns, accepting `AggregateFn`, which are not bound to a column, makes more sense to me **if our goal is to try to match the Pandas `.agg()` API**.

Reply:
Yea, we should accept what you are accepting now. I think my question is do we need to accept `Aggregation` as well? What if user wants to use a custom `on` fn instead of using the column name, how can they do that?

Reply:
@jjyao I'm extending the dict keys to also allow callables, so custom data subsetting should be doable via that.